### ISSUE TITLE
Inconsistent computation of accuracy with lightning

### ISSUE THREAD

**Issue by ashleve** [PyTorchLightning/metrics#597]
Hello, I have the following code:
```python
from torchmetrics.classification.accuracy import Accuracy

class LitModel(LightningModule):
  def __init__(...):
      ...
      self.val_accuracy = Accuracy()
  
  def validation_step(self, batch: Any, batch_idx: int):
      ...

      acc = self.val_accuracy(preds, targets)
      self.log("val/acc", acc, on_step=False, on_epoch=True, prog_bar=True)
  
      return {"loss": loss, "preds": preds, "targets": targets}
  
  def validation_epoch_end(self, outputs: List[Any]):
      acc = self.val_accuracy.compute()
      self.log("val/acc_2", acc, prog_bar=True)
```
So the above code computes accuracy in 2 ways. What bothers me is how different are those 2 accuracies, even though, as far as I understand it, those values should be exactly the same. Here's what I get:
![image](https://user-images.githubusercontent.com/43458894/139592828-84bb70a5-3d39-487f-80ed-3bc952ce0651.png)
(here the difference for MNIST classification is almost 1%)

What is the reason for this difference? Am I using the metrics in a wrong way?

**Comment by SkafteNicki** [comment]
Hi @ashleve,
The correct way to compute the value is the way you compute `val/acc_2`. The reason `val/acc` probably is wrong, is due to the accuracy calculated as the average over the accuracy per batch is not nessesarily the same as the global accuracy. It only holds if the batch size is the same for all batches (it will often be smaller for the last batch).
See the below example showcasing this difference:
```python
from torchmetrics import Accuracy
import torch

metric = Accuracy(num_classes=5)
# batch size of 50
val1 = metric(torch.randn(50,5).softmax(dim=-1), torch.randint(5, (50,)))
# batch size of 20
val2 = metric(torch.randn(20,5).softmax(dim=-1), torch.randint(5, (20,)))

print((val1 + val2) / 2.0)  #tensor(0.1900)
print(metric.compute())  # tensor(0.1857)
```

**Comment by Borda** [comment]
@SkafteNicki do you have a suggestion that we shall be more memory hungry and save all intermediate measures?

**Comment by ashleve** [comment]
@SkafteNicki Thanks for the reply,

Am I correct to assume that setting `drop_last=True` on validation dataloader should make the two accuracies exactly the same? I just did that and still experience a significant difference.

I can prepare a reproducible example if you want

**Comment by SkafteNicki** [comment]
@ashleve yes, that should be the case.
Please send me an reproduceable example so I can get to the bottom of this.

**Comment by ashleve** [comment]
@SkafteNicki Here's the MNIST example:
https://github.com/ashleve/lightning-metrics-bug

**Comment by SkafteNicki** [comment]
Hi @ashleve,
I took a look at your code and figured out the problem was something else. You basically need to change your `validation_epoch_end` to the following:
```python
def validation_epoch_end(self, outputs):
    acc = self.val_acc.compute()  # get val accuracy from current epoch       
    self.log("val/acc_2", acc, prog_bar=True)
    self.val_acc.reset()
```
you need to reset the metric in between epochs. After looking at the documentation, this does not seem to be super clear and I will try to improve on that. The reason the missing reset was messing with your results was due to *Validation sanity check* that runs before the fit. Essentially the one metric was only computing on the validation set whereas the other was computation on the hole validation set + 2 extra batches from the validation sanity checking.

**Comment by ashleve** [comment]
@SkafteNicki how about making lightning reset all metrics after validation sanity checks? Do you think it makes sense, or would it break something?

**Comment by XiaomoWu** [comment]
Thanks for the clarification @SkafteNicki, missing `.reset()` is the culprit. I didn't realize that I should log directly instead of logging the computed value. My ExplainedVariance almost doubles after adding `.reset()`! However, this is such a painful lesson for me to learn. Issue #269 also discusses this. 

Thank you for improving the document. I think it's super important for future users.

**Comment by SkafteNicki** [comment]
@ashleve we could do that, however you would still have to do manual resetting between epochs e.g. calling `metric.reset()` yourself because you do not want to accumulate between epochs, which will also solve the problem with validation sanity checking. I still think the better option better documentation.

**Comment by ashleve** [comment]
@SkafteNicki I'm a bit confused, doesn't lightning call `.reset()` for every metric at the end of every epoch? I believe this is [stated in the documentation](https://torchmetrics.readthedocs.io/en/latest/pages/lightning.html).

**Comment by ashleve** [comment]
@SkafteNicki Thanks for clarifying, I didn't realise `.reset()` is not called automatically at the end of each epoch

**Comment by SkafteNicki** [comment]
@ashleve to be fair we did automatically reset it at the end of each epoch a couple of releases ago, but we changed it since many users wanted more flexibility in when their metrics are reset.

**Comment by SkafteNicki** [comment]
Fixed by https://github.com/PyTorchLightning/metrics/pull/615. Closing.
